core/src/com/fallenflame/game/AIController.java

```diff
@@ -1,5 +1,411 @@
+        ticks = 0;
+
+        random = new Random(randomSeed);
+        randomID = random.nextInt();
+    }
+
+    /**
+     * Returns the action selected by this InputController
+     *
+     * The returned int is a bit-vector of more than one possible input
+     * option. This is why we do not use an enumeration of Control Codes;
+     * Java does not (nicely) provide bitwise operation support for enums.
+     *
+     * This function tests the environment and uses the FSM to chose the next
+     * action of the enemy. This function SHOULD NOT need to be modified.  It
+     * just contains code that drives the functions that you need to implement.
+     *
+     * @return the action selected by this InputController
+     */
+    public Action getAction() {
+        ticks++;
+
+        if ((randomID + ticks) % 10 == 0) {
+            changeStateIfApplicable();
+
+            markGoalTiles();
+            move = getMoveAlongPathToGoalTile();
+        }
+
+        Action action = move;
+        return action;
+    }
+
+    /**
+     * Change the state of the enemy using a Finite State Machine.
+     */
+    private void changeStateIfApplicable() {
+        int rand_int = random.nextInt(100);
+        switch (state) {
+            case SPAWN:
+                state = FSMState.WANDER;
+                break;
+
+
+            case WANDER:
+                if(withinChase()){
+                    state = FSMState.CHASE;
+                    break;
+                }
+
+                FlareModel closestFlare = null; //Safety for if closestFlare is not null for some-reason.
+                //This is to prevent an enemy from getting stuck between two flares
+                int closestFlareDistance = FLARE_DETECTION_RADIUS;
+                for(FlareModel flare : flares){
+                    double flareDistance = cartesianDistance(
+                            level.screenToTile(flare.getX()),
+                            level.screenToTile(enemy.getX()),
+                            level.screenToTile(flare.getY()),
+                            level.screenToTile(enemy.getY())
+                    );
+
+                    if(flareDistance <= closestFlareDistance) {
+                        closestFlare = flare;
+                        closestFlareDistance = (int) flareDistance;
+                    }
+                }
+
+                if(closestFlare != null){
+                    investigatedFlares.add(closestFlare);
+                    investigationPosition = new Vector2(closestFlare.getX(), closestFlare.getY());
+                    state = FSMState.INVESTIAGE;
+                }
+
+                break;
+
+            case CHASE:
+                enemy.setActivated(true);
+
+                if(rand_int < 30 || !withinChase()){ //random chance of quitting chase
+                    state = FSMState.INVESTIAGE;
+                    investigationPosition = new Vector2(player.getX(), player.getY());
+                }
```

Comment:
Also how is setting player's position as investigate position and then investigating different than chasing them?

Reply:
investigationPosition is for the INVESTIGATE state. Chasing updates every single time an enemy is chasing a player, but investigationPosition is set when the player transitions into INVESTIGATE. Once an enemy is in INVESTIGATE, it receives no further information about the player's position. This is how I implemented the enemy going to the player's last-known location

Reply:
As for the random quitting chasing, I think that's a discussion to be had but I've removed it from now. It's basically to prevent the enemies from being near-certain kill-bots and to give the player a chance at evading them in a chase, especially being that enemies are faster than players